notes.md

```diff
@@ -0,0 +1,255 @@
+
+I think what is proposed should be doable without making breaking changes since it just involves a changing the output types of some rules, which isn't something that we consider breaking provided that they represent the same thing. I'd prefer we worry about this later if we think this is a good idea though.
+
+Apologies in advance for the lenght. I've tried to condense where possible, but there's a decent amount of content to get through.
+
+
+
+
+
+## Cutting to the Chase
+
+Under the proposed system, rule-implementers would write rules that look like this:
+```julia
+function rrule(config::RuleConfig, ::typeof(*), A::AbstractMatrix, B::AbstractMatrix)
+
+    # Do the primal computation.
+    C = A * B
+
+    # "natural pullback": write intuitive pullback, closing over stuff in the usual manner.
+    function natural_pullback_for_mul(C̄_natural)
+        Ā_natural = C̄_natural * B'
+        B̄_natural = A' * C̄_natural
+        return NoTangent(), Ā_natural, B̄_natural
+    end
+
+    # Make a call to utility functionality which transforms cotangents of C, A, and B.
+    # Rule-writing without this utility has similar requirements to `ProjectTo`.
+    return C, wrap_natural_pullback(config, natural_pullback_for_mul, C, A, B)
+end
+```
+I'm proposing to coin the term "natural pullback" for pullbacks written within this system, as they're rules written involving natural (co)tangents.
+
+Authors will have to implement two functions for their `AbstractArray` type `P`:
+```julia
+pullback_of_destructure(::P)
+pullback_of_restructure(::P)
+```
+which are the pullbacks of two functions, `destructure` and `(::Restructure)`, that we'll define later.
+
+The proposed candidates for natural (co)tangents are obtained as follows:
+1. natural tangents are obtained from structural tangents via the pushforward of `destructure`,
+2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.
+
+Our current `ProjectTo` functionality is roughly the same as the pullback of `destructure`.
+
+In the proposed system, natural (co)tangents remain confined to `rrule`s, and rule authors can choose to work with either natural, structural, or a mixture of (co)tangents.
+
+Other than the headlines at the top, additional benefits of (a correct implementation of) this approach include:
+1. no chance of trying to sum tangents failing because one is natural and the other structural,
+1. no risk of obstructing AD,
+1. rand_tangent can just use structural tangents, simplifying its implementation and improving its reliability,
+1. we can probably finally make `to_vec` treat things structurally (although we also need to extend it in other ways), which will also deal with reliability / simplicity of implementation problems,
+1. generic constructor for composite types easy to implement,
```

Comment:
Precisely. Does the following make things clearer:
```suggestion
1. generic constructor for composite types easy to implement since we can be sure of obtaining a structural tangent, and rid ourselves of the "need to defin an adjoint for constructor..." errors we see in Zygote,
```